compartmentalize-grains-via-pop.md

```diff
@@ -0,0 +1,71 @@
+- Feature Name: compartmentalize-grains-via pop
+- Start Date: 2020-06-22
+- SEP Status: Draft
+- SEP PR: (leave this empty)
+- Salt Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Grains is currently a static component of salt.  We want it to be developed in a more compartmentalized fashion.
+The engineers at saltstack have created kernel-specific projects that functionally replace the grains
+provided by salt with POP.  We propose that this project, codenamed `corn` replaces grains in salt while
+maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+"The original implementation of grains was not designed to scale to the level that salt requires" - thatch45
+Tom wrote POP to solve this problem and we're going to use it. 
+
+By using Plugin Oriented Programming, grains can be separated into individual projects that can be maintained
+by the community safely.
+
+This also solves the issue of `core.py` being an unmaintainable monolith.  It allows platform specific grains
+to be independently maintained.
+
+Working group leaders and other community members focused on specific platforms can manage the release cycle for
+individual grains projects independently from salt.
+
+# Design
+[design]: #detailed-design
+
+Grains are currently created by calling functions in the salt loader.  We would simply need to call
+external POP functions to collect grains.  We need to maintain backwards compatibility with custom user grains.
+
+- All salt grains have already been ported to and tested on the new platform.
+- The new platform has many more features than the salt-grains implementation, but is compatible with salt grains.
```

Comment:
Will the current methods of setting grains be supported in corn? 

1. `_grains` directory
2. grains in minion config
3. grains in `/etc/salt/grains` & the grain modules which modify grains

Reply:
Yes, grains will be able to be added to salt the same ways they always have.   
You will also be able to add grains the new pop/idem way (such as pip installing a project like [corn cowsay](https://pypi.org/project/corn-cowsay) into the same python environment as salt.

Reply:
A couple of additional compatibility questions:

1. What about other ways to add custom grains (other than `_grains`): `salt-call --module-dirs DIR`, [`grains_dirs`](https://github.com/saltstack/salt/issues/32951) setting, [setuptools entry points](https://docs.saltstack.com/en/latest/topics/tutorials/packaging_modules.html#external-modules-setuptools-entry-points-support)?

2. Will a custom grain that accepts the optional `grains` argument get the core grains passed in? See this feature that was introduced in Salt Fluorine: https://github.com/saltstack/salt/pull/47372

Reply:
Also, is there a plan to sync corn grains with the recent changes in Salt core grains? For example, the recent FQDNs grain change and the new setting that disables it on some platforms https://github.com/saltstack/salt/pull/55581

Maybe a core grain code freeze period could help?

Reply:
Yes, a core grain code freeze is part of this SEP.  We don't want to maintain two forks of grains and all future development of grains should go into the compartmentalized projects..

Other ways of adding grains will remain available as they are.  

A custom grain that accepts the optional `grains` argument would get the grains generated by grainsv2.

Yes, the test in the [Grainsv2 PR](https://github.com/saltstack/salt/pull/57681) shows a line by line diff of grainsv2 vs salt grains